packages/core/echo/echo/src/experimental/query-2.test.ts

```diff
@@ -0,0 +1,58 @@
+import { Schema } from 'effect';
+import { Type } from '..';
+
+declare const Relation: any;
+declare const Query: any;
+
+//
+// Example schema
+//
+
+const Person = Schema.Struct({
+  name: Schema.String,
+}).pipe(Type.def({ typename: 'dxos.org/type/Person', version: '0.1.0' }));
+
+const Org = Schema.Struct({
+  name: Schema.String,
+}).pipe(Type.def({ typename: 'dxos.org/type/Org', version: '0.1.0' }));
+
+const WorksFor = Schema.Struct({
+  since: Schema.String,
+}).pipe(Relation.def({ typename: 'dxos.org/type/WorksFor', version: '0.1.0', source: Person, target: Org }));
+
+const Task = Schema.Struct({
+  title: Schema.String,
+  createdAt: Schema.String,
+  assignee: Type.Ref(Person),
+}).pipe(Type.def({ typename: 'dxos.org/type/Task', version: '0.1.0' }));
+
+//
+// Example queries
+//
+
+// Query<Person>
+const getAllPeople = Query.type(Person);
+
+// Query<Person>
+const getAllPeopleNamedFred = Query.type(Person, { name: 'Fred' });
+
+// Query<Org>
+declare const fred: any;
+const getAllOrgsFredWorkedForSince2020 = Query.type(Person, { id: fred.id })
+  .outgoingRelations(WorksFor, { since: Query.gt('2020') })
+  .targets();
+
```

Comment:
Consider returning a sub-graph rather than chaining successive filters and traversals. E.g., return employees and portfolio companies of venture companies that Fred worked for?

Prior art:

```
d3
  .select()
  .select()
  .call(sel => sel.select()) // Branch
  .select()
```

Reply:
you can do that with a union, and since queries are immutable you can reuse subqueries:

```ts
const ventureCompaniesFredWorkedFor = Query.type(Person, { id: fred.id }).sourceOf(VentureCompany);

const resultQuery = Query.all(
   ...employeeQuery,
  ...portfolioCompanyQuery,
)
```

Reply:
I don't think d3 applies here since that's an imperative API, but change my mind